src/Autofac/Features/ResolveAnything/AnyConcreteTypeNotAlreadyRegisteredSource.cs

```diff
@@ -91,16 +91,39 @@ public AnyConcreteTypeNotAlreadyRegisteredSource(Func<Type, bool> predicate)
                 return Enumerable.Empty<IComponentRegistration>();
             }
 
+            // Check to resolve issue #925
             if (typeInfo.IsGenericType)
             {
-                foreach (var typeParameter in typeInfo.GenericTypeArguments)
+                if (typeInfo.GetGenericTypeDefinition() == typeof(Lazy<>))
                 {
-                    // Issue #855: If the generic type argument doesn't match the filter don't look for a registration.
-                    if (_predicate(typeParameter) && !registrationAccessor(new TypedService(typeParameter)).Any())
+                    return Enumerable.Empty<IComponentRegistration>();
+                }
+
+                var typeParameters = typeInfo.GenericTypeArguments
+                    .Where(t => t.IsAbstract &&
+                        !registrationAccessor(new TypedService(t)).Any());
+
+                bool isValidConstructorFound = false;
+                foreach (var constructor in typeInfo.GetConstructors())
+                {
+                    var parameterTypes = constructor.GetParameters()
+                        .Select(p => p.ParameterType);
+
+                    var isTypeParameterInConstructor = typeParameters
+                        .Where(t => parameterTypes.Contains(t))
```

Comment:
When `parameterTypes.Contains` is called, does it re-run the `Select` above? (I have a feeling it might.) In which case, maybe there should be, say, a `ToArray()` above to force it to evaluate once and be done?

Reply:
Good point, I'll add `ToArray()` like you said.

Reply:
Minor perf point, ToList will be more efficient than ToArray in this case.